[HLSL] select scalar overloads for vector conditions

llvm-beanz opened the pull request:
This PR adds scalar/vector overloads for vector conditions to the `select` builtin, and updates the sema checking and codegen to allow scalars to extend to vectors.

Fixes #126570

llvm-beanz committed:
[HLSL] select scalar overloads for vector conditions

This PR adds scalar/vector overloads for vector conditions to the
`select` builtin, and updates the sema checking and codegen to allow
scalars to extend to vectors.

Fixes #126570

clang-format
clang-format
'cbieneman/select' on '44f0fe9a2806'.

spall commented on clang/test/SemaHLSL/BuiltIns/select-errors.hlsl:
```diff
@@ -1,119 +1,65 @@
-  // viable: requires 3 arguments, but 4 were provided}}
-}
 
 int test_select_first_arg_wrong_type(int1 p0, int t0, int f0) {
```
This and the 2 tests below don't error? Do you just want to delete them?

spall commented on clang/test/SemaHLSL/BuiltIns/select-errors.hlsl:
```diff
@@ -1,119 +1,65 @@
 }
 
 // if a vector second arg isnt a vector
 int2 test_select_builtin_second_arg_not_vector(bool2 p0, int t0, int2 f0) {
```
same here and the one below.

llvm-beanz committed:
Added comments to address @spall's feedback and clang-format

llvm-beanz committed:
Alother slight suffling of hlsl includes

This fixes issues that were causing the new select overloads to fail to
compile.

farzonl commented on clang/lib/Headers/hlsl/hlsl_intrinsic_helpers.h:
```diff
@@ -0,0 +1,71 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _HLSL_HLSL_INTRINSIC_HELPERS_H_
```
How does the `enable_if_t` template in  `hlsl_detail.h`get exposed here?

llvm-beanz replied:
hlsl_detail.h is included first in hlsl.h, which makes it included before the other headers. These headers are all implementation details, and aren't expected to be exposed to user code since hlsl.h is implicitly included in all HLSL source files.

llvm-beanz committed:
NFC. Clang-format

llvm-beanz committed:
Make clang-format happy and don't break the include order

llvm-beanz committed:
Adding some comments
